### ISSUE TITLE
Implement 3x3 double transpose using avx512

### ISSUE THREAD

**Issue by romeric** [romeric/Fastor#55]
Using `_mm512_permutevar_pd`. Something like

~~~c++
//             0,1,2,
//             3,4,5,
//             6,7,8
       __m512d trans07 = _mm512_loadu_pd(src);
        const __m512i trans_mask = _mm512_setr_epi64(
             0,3,6,
             1,4,7,
             2,5);
        __m512d out = _mm512_permutevar_pd(trans07, trans_mask);
       _mm512_storeu_pd(dst,out);
       dst[8] = src[8];
~~~

Load and store aside, this optimises to one cycle! Note that `_mm512_permutevar_pd` might not permute across 128/256 boundaries

**Comment by romeric** [comment]
For the record implementing the 3x3 float case using a similar technique with AVX2:

SSE2 version used currently
~~~c++
    __m128 row0 = _mm_load_ps(a);
    __m128 row1 = _mm_loadu_ps(a+3);
    __m128 row2 = _mm_loadu_ps(a+6);

    __m128 T0 = _mm_unpacklo_ps(row0,row1);
    __m128 T1 = _mm_unpackhi_ps(row0,row1);

    row0 = _mm_movelh_ps ( T0,row2 );
    row1 = _mm_shuffle_ps( T0,row2, _MM_SHUFFLE(3,1,3,2) );
    row2 = _mm_shuffle_ps( T1,row2, _MM_SHUFFLE(3,2,1,0) );

    _mm_store_ps(out,row0);
    _mm_storeu_ps(out+3,row1);
    _mm_storeu_ps(out+6,row2);
~~~

GCC 9.2 generated asm on macOS

~~~asm
	vmovups	12(%rsp), %xmm3
	vmovaps	(%rsp), %xmm0
	vmovups	24(%rsp), %xmm2
	vunpcklps	%xmm3, %xmm0, %xmm1
	vunpckhps	%xmm3, %xmm0, %xmm0
	vmovlhps	%xmm2, %xmm1, %xmm3
	vmovaps	%xmm3, 64(%rsp)
	vshufps	$222, %xmm2, %xmm1, %xmm1
	vshufps	$228, %xmm2, %xmm0, %xmm0
	vmovups	%xmm1, 76(%rsp)
	vmovups	%xmm0, 88(%rsp)
~~~

AVX2 version

~~~c++
       __m256 trans07 = _mm256_loadu_ps(a);
        const __m256i trans_mask = _mm256_setr_epi32(
             0,3,6,
             1,4,7,
             2,5);
        // does not shuffle across 256 lanes, only 128 lanes
        // __m256 _res = _mm256_permutevar_ps(trans07, trans_mask);
        // this one shuffles correctly
        __m256 _res = _mm256_permutevar8x32_ps(trans07, trans_mask);
       _mm256_storeu_ps(out,_res);
       out[8] = a[8];
~~~

GCC 9.2 on macOS gives

~~~asm
	vmovdqa	lC2(%rip), %ymm0
	vpermps	(%rsp), %ymm0, %ymm0
	vmovaps	%ymm0, 64(%rsp)
	vmovss	32(%rsp), %xmm0
	vmovss	%xmm0, 96(%rsp)
~~~

The latter one is just one cycle as opposed to 5 cycles of SSE2 version but `vpermps` is being performed on the memory which is concerning.

**Comment by romeric** [comment]
For the record for the above AVX2 version clang on macos emits this

SSE2
~~~asm
	vmovaps	96(%rsp), %xmm0
	vmovups	108(%rsp), %xmm1
	vmovups	120(%rsp), %xmm2
	vunpcklps	%xmm1, %xmm0, %xmm3 ## xmm3 = xmm0[0],xmm1[0],xmm0[1],xmm1[1]
	vunpckhps	%xmm1, %xmm0, %xmm0 ## xmm0 = xmm0[2],xmm1[2],xmm0[3],xmm1[3]
	vmovlhps	%xmm2, %xmm3, %xmm1 ## xmm1 = xmm3[0],xmm2[0]
	vshufps	$222, %xmm2, %xmm3, %xmm3 ## xmm3 = xmm3[2,3],xmm2[1,3]
	vmovaps	%xmm1, 32(%rsp)
	vblendps	$3, %xmm0, %xmm2, %xmm0 ## xmm0 = xmm0[0,1],xmm2[2,3]
	vmovups	%xmm3, 44(%rsp)
	vmovups	%xmm0, 56(%rsp)
~~~
AVX2
~~~asm
	vmovaps	LCPI0_1(%rip), %ymm0    ## ymm0 = [0,3,6,1,4,7,2,5]
	vpermps	96(%rsp), %ymm0, %ymm0
	vmovaps	%ymm0, 32(%rsp)
	movl	128(%rsp), %eax
	movl	%eax, 64(%rsp)
~~~
It is hard to measure the performance at this level, but the SSE version looks better to me

**Comment by romeric** [comment]
For the following 3 implementations (SSE/AVX/AVX512)

~~~c++
inline void _transpose1(const double* a, double* out) {
    __m128d a11 = _mm_loadu_pd(a);
    __m128d a12 = _mm_loadu_pd(a+2);
    __m128d a21 = _mm_loadu_pd(a+4);
    __m128d a22 = _mm_loadu_pd(a+6);

    _mm_storeu_pd(out  ,_mm_shuffle_pd(a11,a12,0x2));
    _mm_storeu_pd(out+2,_mm_shuffle_pd(a22,a11,0x2));
    _mm_storeu_pd(out+4,_mm_shuffle_pd(a21,a22,0x2));
    _mm_storeu_pd(out+6,_mm_shuffle_pd(a12,a21,0x2));
    _mm_store_sd (out+8,_mm_load_sd(a+8));
}
~~~

~~~c++
inline void _transpose2(const double* a, double* out) {
   __m256d row1 = _mm256_loadu_pd(a);
   __m256d row2 = _mm256_loadu_pd(a+4);

   __m128d a11 = _mm256_castpd256_pd128(row1);
   __m128d a12 = _mm256_extractf128_pd(row1,0x1);
   __m128d a21 = _mm256_castpd256_pd128(row2);
   __m128d a22 = _mm256_extractf128_pd(row2,0x1);

   row1 = _mm256_castpd128_pd256(_mm_shuffle_pd(a11,a12,0x2));
   row1 = _mm256_insertf128_pd(row1,_mm_shuffle_pd(a22,a11,0x2),0x1);
   row2 = _mm256_castpd128_pd256(_mm_shuffle_pd(a21,a22,0x2));
   row2 = _mm256_insertf128_pd(row2,_mm_shuffle_pd(a12,a21,0x2),0x1);

   _mm256_storeu_pd(out,row1);
   _mm256_storeu_pd(out+4,row2);
   _mm_store_sd(out+8,_mm_load_sd(a+8));
}
~~~

~~~c++
inline void _transpose3(const double* a, double* out) {
    __m512d a07 = _mm512_loadu_pd(a);
    const __m512i trans_mask = _mm512_setr_epi64(
        0,3,6,
        1,4,7,
        2,5);
    __m512d trans07 = _mm512_permutexvar_pd(trans_mask, a07);
    _mm512_storeu_pd(out,trans07);
    _mm_store_sd(out+8,_mm_load_sd(a+8));
}
~~~

godbolt gives the following time (looping over all of them 1Million times) [SSE/AVX/AVX512]

GCC 9.3
~~~
23 runs, min time: 21.6551 ms. mean time: 22.9657 ms. max time: 29.8916 ms. Average no of RDTSC CPU cycles 45246359
27 runs, min time: 17.1194 ms. mean time: 19.1362 ms. max time: 24.0902 ms. Average no of RDTSC CPU cycles 37701589
32 runs, min time: 14.7661 ms. mean time: 15.8746 ms. max time: 22.0919 ms. Average no of RDTSC CPU cycles 31275554
~~~

Clang 10.0.0
~~~
25 runs, min time: 20.8001 ms. mean time: 20.0817 ms. max time: 21.1479 ms. Average no of RDTSC CPU cycles 39564380
31 runs, min time: 16.5846 ms. mean time: 16.6307 ms. max time: 19.5026 ms. Average no of RDTSC CPU cycles 32765310
35 runs, min time: 14.3441 ms. mean time: 14.5738 ms. max time: 17.9571 ms. Average no of RDTSC CPU cycles 28712755
~~~

ICC 19.0.1
~~~
18 runs, min time: 29.6749 ms. mean time: 28.0493 ms. max time: 29.7266 ms. Average no of RDTSC CPU cycles 55261854
16 runs, min time: 33.9149 ms. mean time: 31.8149 ms. max time: 33.958 ms. Average no of RDTSC CPU cycles 62680816
26 runs, min time: 19.6943 ms. mean time: 19.8985 ms. max time: 22.9203 ms. Average no of RDTSC CPU cycles 39203383
~~~

So clearly AVX512 version is the winner.

**Comment by romeric** [comment]
This is now implemented 9fa74364db355d8c7ffabf217c53cdfdd2854c8e. Closing.
